### DISCUSSION TITLE
NFS client directory doesn't match directory on mergerfs host

### DISCUSSION THREAD

**Discussion by hwwilliams** [trapexit/mergerfs#1183] [General]
I have noticed that certain directories on NFS clients connected to the mergerfs pool are accessing a directory that is not up to date with the mergerfs pool on the mergerfs host.

The mergerfs host is configured for NFS v4 only and the client is configured to mount the NFS share (mergerfs pool) using the following options: 10.0.0.10:/comics /mnt/mergerfs/comics nfs defaults,nfsvers=4,minorversion=2,proto=tcp,port=2049 0 0

I have attached a screenshot below of an NFS client on the top where a comic issue 001 is missing but the mergerfs host does have the issue 001 in the same directory.

![image](https://github.com/trapexit/mergerfs/assets/20447447/5d24378e-bafd-46bd-a39e-c8d005910337)

I'm not doing any file caching but clearly something is being cached or not updated, does anyone have any suggestions?

OS: Debian 11
Mergerfs Version: 2.35.1
Mergerfs Mount Options: /mnt/mergerfs/data* /mnt/mergerfs/pool fuse.mergerfs branches-mount-timeout=300,cache.files=off,category.create=pfrd,dropcacheonclose=true,fsname=mergerfs,inodecalc=path-hash,lazy-umount-mountpoint=true,link_cow=true,noforget,readahead=2048,threads=16,xattr=nosys 0 0

**Comment by trapexit** [comment]
Are you changing things out of band to mergerfs or nfs?

**Reply by hwwilliams** [reply]
If you mean manipulating files on the disks themselves, no. All file access is done directly to the pooled directory and 99% of it is over NFS, if its not over NFS then its me doing something on the host but its still under the pooled directory.

**Reply by trapexit** [reply]
What you describe sounds contradictory. 

You have?: underlying fs <-> mergerfs <-> nfs export

correct?

If so and you are changing anything at the mergerfs or underlying filesystem level that's not good. NFS isn't going to like that. All changes should be happening through NFS.

**Reply by hwwilliams** [reply]
> underlying fs <-> mergerfs <-> nfs export

Underlying file system being the normal disks that I have mounted at /mnt/mergerfs/data* then no I never touch that.

> NFS isn't going to like that. All changes should be happening through NFS.

Alright fair, I didn't realize that but in this case the file that is not showing up isn't something that I've touched at the mergerfs level either. The file was created by one client (Mylar) over NFS and another client (Kavita) is trying to read it over NFS.

**Reply by trapexit** [reply]
If you are seeing the file in mergerfs and set things up so as to not upset NFS, which you have, it must have something to do with NFS caching and there isn't much I can comment on without a minimal reproducible example and more details as described in the support section of the docs or the issue ticket template here on github. There are a lot of moving parts here and nothing can be ignored or taken for granted.

**Reply by hwwilliams** [reply]
Well I restarted the Kavita VM, which obviously remounted the share and cleared any kind of cache, and the files show up fine now but all the docs I'm reading say that NFS caching is very low duration by default with a maximum of only 60 seconds depending on which option you're looking at but this file was missing for way longer than 60 seconds in that directory on the Kavita client so I'm still confused if this was just a NFS caching issue.

**Reply by hwwilliams** [reply]
I have found this document titled [Why NFS Sucks](https://www.kernel.org/doc/ols/2006/ols2006v2-pages-59-72.pdf) which is an interesting read so far, sounds like I either have to live with inconsistent cache or lower the cache amount which apparently kills performance. Makes me just want to find another way to access the files over the network.

Thanks for the help narrowing it down to NFS as mergerfs is an even bigger mystery to me.

**Reply by trapexit** [reply]
If mergerfs is listing the files then it is likely beyond mergerfs which would mean the kernel, NFS server, or NFS client. mergerfs is mostly just a proxy. The kernel asks questions and it responds. If you enable certain caching pretty much all of it is kernel side. Not mergerfs. The only cache that is happening in mergerfs is statfs cache which has no play here. At most something is happening to confuse NFS but as mentioned in the docs that would typically be inode related and result in stale files.

You can easily test this by stracing mergerfs when this stuff is happening. Reproduce the issue, make sure nothing else is using mergerfs/nfs/etc., start stracing mergerfs as described in the docs, run "ls /problematic/dir/" on the NFS client. Then we can look to see what requests mergerfs gets and the responses.

**Reply by trapexit** [reply]
> I have found this document titled [Why NFS Sucks](https://www.kernel.org/doc/ols/2006/ols2006v2-pages-59-72.pdf) which is an interesting read so far, 

Standard *unix filesystem APIs just weren't designed to be distributed and any solution will either have to make some compromise. Whether that is to the standards in hopes of getting better perf or keeping with standards and losing perf. They are super useful but can have issues. Samba / CIFS is more forgiving to my understanding to some of these things and I believe can be setup with posix acls but I've always found the documentation on it (and NFS for that matter) terrible. Depending on your needs there is that, sshfs (which can be tweaked for some better perf but still not great), or 9pfs but the latter hasn't had much support from what I've seen in a number of years. I've thought about trying to build a simple network filesystem that doesn't try to do anything fancy but I've just not had the time.

I'd love to have a better answer for people looking for remote filesystems and maybe I'm just missing a solution. If someone comes up with a more fool proof setup (config and all) I'd be happy to document it for others.

**Reply by hwwilliams** [reply]
> You can easily test this by stracing mergerfs when this stuff is happening. Reproduce the issue, make sure nothing else is using mergerfs/nfs/etc., start stracing mergerfs as described in the docs, run "ls /problematic/dir/" on the NFS client. Then we can look to see what requests mergerfs gets and the responses.

I'll have to wait for it to happen again on the comics export or see if any of the other exports are having the same issue, if it pops up again I'll try this and see what I get, I didn't want to leave this open and try to make this seem like a mergerfs issue.
